Simplenote/build.gradle

```diff
@@ -36,7 +37,13 @@ android {
         minSdkVersion 23
         targetSdkVersion 29
 
-        testInstrumentationRunner 'androidx.test.runner.AndroidJUnitRunner'
+        testInstrumentationRunner 'com.automattic.simplenote.SimplenoteAppRunner'
```

Comment:
Warning (⚠️): By changing the `testInstrumentationRunner`, all current tests, that require the default `AndroidJUnitRunner` runner, will be starting to fail with the below stack trace, which is unrelated to having a valid account defined in `gradle.properties`. Have you considered an approach where multiple test instrumentation runners can be used.

We need to make sure that the already existing instrumentation tests are unaffected by the new `integration` ones? Otherwise running those locally using the CLI or remotely on CI will stop working (`./gradlew connectedDebugAndroidTest`). Isn't that right?

```
com.automattic.simplenote.NoteListFragmentTest > testNonSearchCursorReturnsAfterSearchApplied[Pixel_4_API_30(AVD) - 11] FAILED 
        java.lang.NullPointerException: Attempt to invoke interface method 'void com.simperium.storage.StorageProvider$BucketStore.reset()' on a null object reference
        at com.simperium.client.Bucket.reset(Bucket.java:902)
[InstrumentationResultParser]: Ignoring repeated failed event for com.automattic.simplenote.NoteListFragmentTest#testNonSearchCursorReturnsAfterSearchApplied. Stack: java.lang.RuntimeException: Unable to resume activity {com.automattic.simplenote.debug/com.automattic.simplenote.NotesActivity}: java.lang.NullPointerException: Attempt to invoke interface method 'void com.simperium.storage.StorageProvider$BucketStore.reset()' on a null object reference
at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4444)
at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:4476)
at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:52)
at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066)
at android.os.Handler.dispatchMessage(Handler.java:106)
at android.os.Looper.loop(Looper.java:223)
at android.app.ActivityThread.main(ActivityThread.java:7656)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
Caused by: java.lang.NullPointerException: Attempt to invoke interface method 'void com.simperium.storage.StorageProvider$BucketStore.reset()' on a null object reference
at com.simperium.client.Bucket.reset(Bucket.java:902)
at com.automattic.simplenote.utils.AuthUtils.logOut(AuthUtils.java:26)
at com.automattic.simplenote.NotesActivity.startLoginActivity(NotesActivity.java:1332)
at com.automattic.simplenote.NotesActivity.onResume(NotesActivity.java:276)
at android.app.Instrumentation.callActivityOnResume(Instrumentation.java:1456)
at androidx.test.runner.MonitoringInstrumentation.callActivityOnResume(MonitoringInstrumentation.java:743)
at android.app.Activity.performResume(Activity.java:8135)
at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4434)
... 11 more
```

Reply:
It is hard to specify different instrumentation class for different tests. We can do that with `Robolectric` (http://robolectric.org/configuring/) but we would need to add the dependency. The problem we are having here is that `NoteListFragmentTest` tries to use our tags bucket and notes buckets and out `TestBucket` does not redefine the method `reset()`. 

One solution we could apply is to have a parameter that we can change to decide whether return the default application bucket (Simperium implementation) or our `Testbucket`. We can specify this parameter at the test level.

Reply:
👋 @dandominguez04 !

> It is hard to specify different instrumentation class for different tests. We can do that with Robolectric (http://robolectric.org/configuring/) but we would need to add the dependency.

I am not sure how `Robolectric` will solve that but I don't mind the extra dependency if it does.

PS: `Robolectric` and `Espresso` as progressively merging together and are now both part of Google, although I must admit that `Robolectric` development is not as active as I would have liked to. However, lately it started to pick up again, so... 🤷 

> The problem we are having here is that NoteListFragmentTest tries to use our tags bucket and notes buckets and out TestBucket does not redefine the method reset().

Yes, exactly that. But, this is not only it, not having the `reset()`, as `TestBucket` cannot be used at all with the existing instrumentation test since they are using the actual live `Bucket` and not mocking anything.

> To solve this problem we are going to define a parameter that we can change to decide whether return the default bucket implementation (Simperium implementation) or our TestBucket. We can specify this parameter at the test level. The default behavior is to return the default implementation. For the tests we will define, we will set this variable to false to use TestBucket objects.

I really like that solution, it is straightforward and doesn't require much build related configuration to what you already did. Seeing what you already do with the `BaseUITest.kt` on your subsequent #1378 PR, this help understanding its usage. Great thinking! 🌟 

```
    @Before
    fun setup() {
        application = ApplicationProvider.getApplicationContext() as SimplenoteTest
        // Make sure to use TestBucket buckets in UI tests
        application.useDefaultBucketImpl = false

        tagsBucket = application.tagsBucket as TestBucket<Tag>
        tagsBucket.clear()
    }
```

The only thing I might have changed, but this is very minor and merely a suggestion, is the `useDefaultBucketImpl` field to `useTestBucket` and revert the logic from `true` to `false`. This is so that to make everything related to the `SimplenoteTest` a test bucket first approach.

Then, a couple of additional refactoring would make this `SimplenoteTest` look like the below:

```
class SimplenoteTest : Simplenote() {
    var useTestBucket = false // Decides whether to use the test bucket implementation or the default.

    private val tagsBucket = object : TestBucket<Tag>("tags") {
        override fun build(key: String?): Tag {
            return Tag(key)
        }
    }

    private val notesBucket = object : TestBucket<Note>("notes") {
        override fun build(key: String?): Note {
            return Note(key)
        }
    }

    override fun getTagsBucket(): Bucket<Tag> = if (useTestBucket) tagsBucket else super.getTagsBucket()

    override fun getNotesBucket(): Bucket<Note> = if (useTestBucket) notesBucket else super.getNotesBucket()
}
```

Then, for the `BaseUITest`, this will make the setup more explicit towards the test buckets instead of the default, since whoever would like to extend from the `BaseUITest` would now know that it actually enables the test buckets instead of disabling the default, like:

```
        application.useTestBucket = true
```

Wdyt? I know naming is hard, so please take all that with a grain of salt and go with whatever you decide is best. 🤝

Reply:
Yes, I think it is simple. I applied the change in commit 62295bd6.

Reply:
👋 @dandominguez04 !

Many thanks, looking good! 🥇 

PS: Thanks for providing me with the commit that applies the change. 🙏